### PR TITLE
boot: add support for tablet hardware buttons (vol up/down, power)

### LINKED ISSUE #30598: Support hardware buttons in systemd-boot
### Component

systemd-boot

### Is your feature request related to a problem? Please describe

Tablets and handheld gaming devices, much like phones, don't typically have physical keyboards.  Therefore, a user cannot select a systemd-boot entry without connecting one.

### Describe the solution you'd like

Following the convention set by Android phones, systemd-boot should support volume up/down in lieu of arrow up/down, and power in lieu of select.  (Tapping power would be used for selection; holding the button down could be used for shutdown).

### Describe alternatives you've considered

It would be really handy if the gamepad buttons on a gaming handheld were also supported; however, I suspect systemd-boot already has plumbing for keyboard events, and it might be out-of-scope to add gamepad support for what's intended to be a minimalist loader.

### The systemd version you checked that didn't have the feature you are asking for

254

### PR DESCRIPTION
**Pull request by appsforartists**
Fixes #30598

### CODE REVIEW

File: src/boot/efi/boot.c [hunk 1]
```diff
@@ -921,6 +923,7 @@ static bool menu_run(
                 case KEYPRESS(0, SCAN_F3, 0): /* EZpad Mini 4s firmware sends malformed events */
                 case KEYPRESS(0, SCAN_F3, '\r'): /* Teclast X98+ II firmware sends malformed events */
                 case KEYPRESS(0, SCAN_RIGHT, 0):
+                case KEYPRESS(0, SCAN_SUSPEND, 0):
```
Comment (poettering) [inline src/boot/efi/boot.c:926]: can you add a comment here, saying this is for covering tablet keys

**Comment by poettering** [comment]
lgtm, just please add one comment, see above.

if this works for you, please drop the "draft" state

**Comment by appsforartists** [comment]
Thanks!

I think I've got the new build installed on my device, but unfortunately the volume keys still don't seem to be doing anything.

**Comment by poettering** [comment]
so the code does make sense to me, but i do not have any hw to test this with.

@appsforartists will you track this down further? (if not we should just close this)

do you get any keypresses at all in uefi? you could just patch things to write the literal code to screen, and see if you can decode it properly to make things work?

**Comment by appsforartists** [comment]
I'm not a C programmer, so I don't know how much farther I can take this without a whole lot of digging.  I was tempted to see if I could figure out how to change what colors it draws on the screen so I can confirm that the modified build was the one being deployed, but I forgot to go do it.

If you wanna bang out a patch that draws detected keycodes somewhere, I'm happy to build it and test it.  I know it's less than ideal to collaborate remotely like that, but it seems that you have the expertise and I have the equipment to test on. 🙃

**Comment by appsforartists** [comment]
I see this section:

https://github.com/systemd/systemd/blob/56afff50b9e0ccfa16c922c2a8f565ec4f1bbda5/src/boot/efi/boot.c#L183-L191

I'm guessing I could add something like `KEYCHAR(key) + ` to `print_at(…, print + cursor)`?

**Comment by poettering** [comment]
you can output numbers with regular C `printf("…");`....

for example `printf("keycode: %" PRIu64 "\n", (uint64_t) key)`

**Pull request closed without merging**